lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -1987,6 +1987,278 @@ LogicalResult ConvertAtenOp<AtenFlattenUsingIntsOp>::matchAndRewrite(
   return success();
 }
 
+template <typename AtenOpT, typename TosaOpT>
+class ConvertAtenPoolingBaseOp : public OpConversionPattern<AtenOpT> {
+public:
+  using OpConversionPattern<AtenOpT>::OpConversionPattern;
+  using OpAdaptor = typename AtenOpT::Adaptor;
+
+  // Different pooling variants need to process inputs differently, e.g.
+  // adaptive pooling generates the kernel size rather than receive it. This
+  // function also transposes inputs.
+  virtual LogicalResult processInputs(AtenOpT op, OpAdaptor adaptor,
+                                      ConversionPatternRewriter &rewriter,
+                                      Value &input, ArrayAttr &kernel,
+                                      ArrayAttr &stride, ArrayAttr &pad,
+                                      Type &outputTy) const {
+    return rewriter.notifyMatchFailure(
+        op, "Unimplemented pooling input parsing function");
+  }
+
+  int64_t getOutputDim(int64_t inputDim, int64_t kernelDim, int64_t stride,
+                       int64_t padBefore, int64_t padAfter,
+                       int64_t dilation) const {
+    if (inputDim == ShapedType::kDynamicSize)
+      return ShapedType::kDynamicSize;
+    else
+      return (
+          (inputDim + padBefore + padAfter - dilation * (kernelDim - 1) - 1) /
+              stride +
+          1);
+  }
+
+  // Apply the transposeDims vector on input to generate a transposed form.
+  Value transposeTensor(AtenOpT op, ConversionPatternRewriter &rewriter,
+                        Value input, SmallVector<int32_t> transposeDims) const {
```

Comment:
it's recommended to use ArrayRef if the list is not modified. refer to https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h